### PR TITLE
Added support for job management through Artemis

### PR DESCRIPTION
**Pull request by Ceiu**
- Added the new JobManager class, which should handle job queuing,
  execution, and other general management tasks
- Added two new fields to JobStatus: job data and job origin
- Added a new method to Util: getHostname
- General refactoring of JobStatus

### CODE REVIEW

File: server/src/main/resources/broker.xml [hunk 1]
```diff
@@ -97,6 +98,20 @@
                 <redelivery-delay>0</redelivery-delay>
                 <max-delivery-attempts>1</max-delivery-attempts>
             </address-setting>
+
+            <address-setting match="event.jobs">
```
Comment (mstead) [inline server/src/main/resources/broker.xml]: Jobs should have its own address and should not be filtered through the 'event'.* path.

File: server/src/main/java/org/candlepin/async/JobManager.java [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        //      JobBuilder builder = new JobBuilder()
+        //          .forTask(runnable job class or key here)
+        //          .setTaskArgument("key", "value")
+        //          .setTaskArgument("key2", "value2")
+        //          .addUniqueRestriction("owner", owner_id_here)
+        //          .addUniqueRestriction("product", product_id_here)
+        //          .addTaskMetadata("correlation_id", cid)
+        //          .setRetryCount(3)
+        //
+        //      jobManager.queueJob(builder);
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/main/java/org/candlepin/async/JobManager.java]: I'm not working on it at the moment (moved to other things in the interim), but I'll gladly start on the JobBuilder now to get Quarts out of the loop.

[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/main/java/org/candlepin/async/JobManager.java]: Looking at it again, I'm still not opposed to ripping out the Quartz stuff and replacing it with our own, but it's big enough that I feel it should be its own PR. We'll need not only the JobBuilder, but some replacement for JobExecutionData, a new interface for the jobs, and then an emulation layer between the quarts stuff and our stuff for the jobs.

It's a pretty sizeable chunk of work. Again, I'm very much in favor of doing so, but not in this PR.

[the remark this answers is not in the document]
Comment (mstead) [reply server/src/main/java/org/candlepin/async/JobManager.java]: I'm cool with doing it in a separate PR to keep the size down, so long as it is still part of the story that is in progress. The definition of done for the story is to build the framework for creating/sending job messages and getting us to the point where we can configure a job, send the message and be ready to start processing them.

I'm not sure that I understand what you mean by 'emulation layer'. The builder collects the data required to configure the job, the JobStatus is created based on this data and persisted, the message is created and sent. What am I missing?

With the necessary bits in place, Quartz/Pinsetter should remain functional until we are ready to start moving the job implementations to the new framework. Ripping the bulk of pinsetter out should be done once this is ready IMO.

**Comment by mstead** [comment]
After some discussion, we've decided to not reuse the current JobStatus, but roll a new version to keep it separate from the existing job framework. The job status resource will temporarily reference both curators when required.

File: server/src/main/java/org/candlepin/model/AsyncJobStatus.java [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+
+    /** Enum of job states; terminal states represent states at which the job will no longer change */
+    public enum JobState {
+        CREATED(false),
+        QUEUED(false),
+        RUNNING(false),
+        COMPLETED(true),
+        FAILED_WITH_RETRY(false),
+        FAILED(true),
+        CANCELED(true);
```
Comment (mstead) [inline server/src/main/java/org/candlepin/model/AsyncJobStatus.java]: Let's add some java doc for each of these describing what they mean.

Comment (Ceiu) [reply server/src/main/java/org/candlepin/model/AsyncJobStatus.java]: Done. I also rearranged the states a bit so the terminal states were at the bottom and the listing is, generally, more in line with the expected state transitions.

**Pull request merged**